components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -195,6 +198,60 @@ class GeckoEngineSession(
         }
     }
 
+    /**
+     * See [EngineSession.requestPdfToDownload]
+     */
+    override fun requestPdfToDownload() {
+        geckoSession.saveAsPdf().then(
+            { inputStream ->
+                if (inputStream == null) {
+                    return@then GeckoResult<Void>()
+                }
+
+                val url = this.currentUrl ?: ""
+                val contentType = "application/pdf"
+                val disposition = currentTitle?.pdfContentDisposition()
+                // A successful status code suffices because the PDF is generated on device.
+                val responseStatus = RESPONSE_CODE_SUCCESS
+                // We do not know the size at this point; send 0 so consumers do not display it.
+                val contentLength = 0L
+                // NB: If the title is an empty string, there is a chance the PDF will not have a name.
```

Comment:
For this case should we just fallback to a default name?

Reply:
We do fallback back to the default name `filename.pdf` but there is a bug in the Downloads code where we get an empty string. A user reported this bug already with mp3/mp4 files, so I felt we should fix that bug separately and referenced it here for now.

Reply:
Make senses 👍🏽